### PR TITLE
[flang][Driver] Support -nostdlib and -nodefaultlibs

### PR DESCRIPTION
**Pull request by tarunprabhu**
This partially addresses some requests in #89888

### CODE REVIEW

**Review by tblah — APPROVED** [review]
LGTM, thanks

File: clang/include/clang/Driver/Options.td [hunk 1]
```diff
@@ -5572,7 +5572,8 @@ def : Flag<["-"], "nocudalib">, Alias<nogpulib>;
 def gpulibc : Flag<["-"], "gpulibc">, Visibility<[ClangOption, CC1Option, FlangOption, FC1Option]>,
   HelpText<"Link the LLVM C Library for GPUs">;
 def nogpulibc : Flag<["-"], "nogpulibc">, Visibility<[ClangOption, CC1Option, FlangOption, FC1Option]>;
-def nodefaultlibs : Flag<["-"], "nodefaultlibs">;
+def nodefaultlibs : Flag<["-"], "nodefaultlibs">,
+  Visibility<[ClangOption, FlangOption, CLOption, DXCOption]>;
```
Comment (Maetveis) [inline clang/include/clang/Driver/Options.td:5576]: @tarunprabhu, @tblah

Shouldn't these be: `Visibility<[ClangOption, FlangOption]>`, it seems like `clang-cl` is now accepting `-nodefaultlibs` without diagnostic, and it does not do anything (does NOT add /NODEFAULTLIB to link.exe's command line)

Comment (tarunprabhu) [reply clang/include/clang/Driver/Options.td:5576]: From what i was able to tell when I changed this, the lack of an explicit visibility implied these 4, but perhaps I was mistaken. I have no way to test clang-cl. Is there a test for this somewhere?

Comment (Maetveis) [reply clang/include/clang/Driver/Options.td:5576]: I don't see anything invoking clang-cl if searching for `-nodefaultlibs` in clang tests: https://github.com/search?q=repo%3Allvm%2Fllvm-project+nodefaultlibs+path%3Aclang%2Ftest&type=code I can check tomorrow (Europe time) if indeed the default visibility extends to clang-cl.

Comment (Maetveis) [reply clang/include/clang/Driver/Options.td:5576]: With this change in Options.td:

```diff
diff --git a/clang/include/clang/Driver/Options.td b/clang/include/clang/Driver/Options.td
index cb96b5daed9d3a2..842e6c2e6233adb 100644
--- a/clang/include/clang/Driver/Options.td
+++ b/clang/include/clang/Driver/Options.td
@@ -5644,4 +5644,3 @@ def gpulibc : Flag<["-"], "gpulibc">, Visibility<[ClangOption, CC1Option, FlangO
 def nogpulibc : Flag<["-"], "nogpulibc">, Visibility<[ClangOption, CC1Option, FlangOption, FC1Option]>;
-def nodefaultlibs : Flag<["-"], "nodefaultlibs">,
-  Visibility<[ClangOption, FlangOption, CLOption, DXCOption]>;
+def nodefaultlibs : Flag<["-"], "nodefaultlibs">;
 def nodriverkitlib : Flag<["-"], "nodriverkitlib">;
 ```
 
I'm getting:
 
```console
> build\bin\clang-cl.exe -nodefaultlibs asd.o
clang-cl: warning: unknown argument ignored in clang-cl: '-nodefaultlibs' [-Wunknown-argument]
```

without it I don't get a warning.

The comment at Options.td:74 also seems to suggest the default is only the gcc compatible driver. https://github.com/llvm/llvm-project/blob/c3536b263f253a69fb336fb0617ee33a01a5c5dd/clang/include/clang/Driver/Options.td#L74-L77

Comment (Maetveis) [reply clang/include/clang/Driver/Options.td:5576]: I opened https://github.com/llvm/llvm-project/pull/118640#issuecomment-2517550992, can you take a look?

Comment (banach-space) [reply clang/include/clang/Driver/Options.td:5576]: Just a drive-by comment.

> From what i was able to tell when I changed this, the lack of an explicit visibility implied these 4, but perhaps I was mistaken.

This should be verifiable by checking Options.inc file in the build directory _before_ and _after_ change. Options.inc is the C++ files generated from Options.td when building `clangDriver`. Please use that as your guidance.

**Pull request merged**